amespahdbpythonsuite/fitted.py

```diff
@@ -10,4 +19,448 @@ class Fitted(Spectrum):
+
+                return None
+
+            classes = self.getclasses(**keywords)
+
+            if not keywords.get('absolute', False):
+                total, err = integrate.quad(self.grid, self.getfit())
+                print('first integrate occurance')
+
+            for key in classes:
+                breakdown[key] = integrate.quad(self.grid, classes[key]) / total
+                print('second integrate occurance')
+
+            return breakdown
+
+        if not self.weights:
+            message('WEIGHTS ARE NOT SET')
+
+            return None
+
+        if not keywords.get('absolute', False):
+            total = np.sum(fweight)
+
+        subclasses = {'anion': {'subclass': 'charge',
+                                'operator': operator.lt,
+                                'operand': 0},
+                      'neutral': {'subclass': 'charge',
+                                  'operator': operator.eq,
+                                  'operand': 0},
+                      'cation': {'subclass': 'charge',
+                                 'operator': operator.gt,
+                                 'operand': 0},
+                      'small': {'subclass': 'nc',
+                                'operator': operator.le,
+                                'operand': keywords.get('small', 50)},
+                      'large': {'subclass': 'nc',
+                                'operator': operator.gt,
+                                'operand': keywords.get('small', 50)},
+                      'nitrogen': {'subclass': 'nn',
+                                   'operator': operator.gt,
+                                   'operand': 0}}
+
+        for key in subclasses:
+            breakdown[key] = self.__breakdown(subclasses[key]) / total
+
+        # Obtaining pure PAH breakdown.
+        uids = [uid for uid in self.uids
+                if self.pahdb['species'][uid]['nn'] == 0
+                and self.pahdb['species'][uid]['no'] == 0
+                and self.pahdb['species'][uid]['nmg'] == 0
+                and self.pahdb['species'][uid]['nsi'] == 0
+                and self.pahdb['species'][uid]['nfe'] == 0]
+
+        if len(uids) > 0:
+            breakdown['pure'] = np.sum([self.weights[uid] for uid in uids]) / total
+
+        # Getting Nc.
+        nc = np.array([self.pahdb['species'][uid]['nc'] for uid in self.uids])
+        breakdown['nc'] = np.sum(nc * fweight) / np.sum(fweight)
+        # Getting fit uncertainty.
+        breakdown['error'] = self.__geterror()
+
+        return breakdown
+
+    def __breakdown(self, s):
+        """
+        Retrieve the sum of the fitting weights for the fitted PAHs.
+
+        """
+        uids = [uid for uid in self.uids
+                if s['operator'](self.pahdb['species'][uid][s['subclass']], s['operand'])]
+
+        if len(uids) > 0:
+            return np.sum([self.weights[uid] for uid in uids])
+
+        else:
+            return 0.0
+
+    def _atoms(self):
```

Comment:
This only needs to be done once; so put perhaps a guard member in place, e.g., self.__hasAtoms, or check if self.pahdb['species'][self.uids[0]]['nc'] exists.

Considering this a bit more, by directly inserting them into self._pahdb['species'] an inconsistent state is created. Since self._pahdb['species'] is the entire database, some entries will have 'nc', 'nh', etc. set, while others won't. So either have a class member, e.g., self._atoms, dictionary that holds that information, which would requires some reworking of __breakdown and __classes, or go all the way back to the XML-parser and add it for all UIDs there.